drivers/sensor/stm32/temp_stm32.h

```diff
@@ -0,0 +1,28 @@
+/*
+ * Copyright (c) 2017 Linaro Limited.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef _STM32_TEMP_H_
+#define _STM32_TEMP_H_
+
+#include <stdint.h>
+
+#include <device.h>
+#include <soc.h>
+
+struct temp_stm32_config {
+	ADC_TypeDef *adc;
+	ADC_Common_TypeDef *adc_common;
+	uint8_t adc_channel;
+#if CONFIG_SOC_SERIES_STM32F4X
+	struct stm32f4x_pclken pclken;
+#endif
```

Comment:
Right, that is usually done when the driver tries to be STM32 generic but only really tested and supported on a specific family, in our case STM32F4X.
Either you restrict everything at Kconfig, or you make it FAMILY_STM32 compatible but protect everything you know that might be MCU specific like you did, but abort when compiling to a different MCU that is not yet tested/supported.

drivers/interrupt_controller/exti_stm32.c is one example ("#error "Unknown STM32 SoC"").